refactor(oracle): introduce FeederManager to replace aggregator and caches for transaction handling

leonz789 opened the pull request:
## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

The current architecture for handling transactions relies on aggregator and cache, which leads to scattered responsibilities, complex state management, and redundant code. To address these issues, this PR introduces a new structure, FeederManager, to streamline the management of caches, token feeders, and KV store state updates.
Refactor oracle module
![image](https://github.com/user-attachments/assets/ab221e69-032d-412f-8830-f7801adc0f6a)

Changes in This PR
1.	FeederManager Implementation:
	-	FeederManager replaces the existing aggregator and cache structures.
	-	It encapsulates the management of caches, token feeders, and KV store state updates.
	
2.	Per-TokenFeeder Round Management:
	-	For each tokenFeeder defined in oracleParams, FeederManager creates a round structure to handle its updates.
	-	The round processes received messages and delegates price-related calculations to an internal aggregator.
	
3.	Caches for Oracle Parameters and Validator Set:
	-	FeederManager maintains caches to sync the latest oracleParams and validatorSet.
	-	It temporarily stores all valid transactions for the current block, writing them to the KV store at EndBlock if updates occur.
	
4.	Message Processing:
	-	When receiving a msg, FeederManager determines the corresponding round based on msg.FeederID and forwards the price data for calculation.
	-	The round delegates message processing to its internal aggregator.
	
5.	Round State Updates:
	-	At EndBlock, each round updates its state (e.g., closing price windows, enabling submissions, or reopening windows).
	-	Depending on the state, the round may write prices to the KV store.
	
6.	Aggregator Price Handling:
	1.	Valid prices are filtered and recorded in aggregator.recordsValidators, grouped by validator.
	2.	For deterministic sources, valid prices are recorded in aggregator.recordsDSs, grouped by sourceID and detID.
	3.	If consensus is reached on a deterministic source price (voting power exceeds the threshold), the price is updated in recordsValidators.
	4.	Non-deterministic source prices are stored as the finalPrice for the source without additional processing.
	5.	Aggregated prices are calculated using aggregator.GetFinalPrice().
	
7.	Price Aggregation Logic:
  The aggregation process is performed in two steps:
    -	Per Validator Aggregation: Compute the finalPrice for each validator by aggregating the finalPrice of all sources associated with the validator in recordsValidators.
    -	Overall Aggregation: Compute the total result by aggregating the finalPrice of all validators.
## Notable changes
- use base64 of hash to emit nst-balance-change-event, don't reset index in when nst rounds grows

----

Closes #XXX

<!--
< < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.

PR review checkboxes:

I have...

- [ ] added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] included the correct
      [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
      in the PR title
- [ ] targeted the correct branch
      (see [PR Targeting](https://github.com/ExocoreNetwork/exocore/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link in the PR description to the relevant issue or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all required CI checks have passed

Code maintenance:

I have...

- [ ] written unit and integration [tests](https://github.com/ExocoreNetwork/exocore/blob/main/CONTRIBUTING.md#testing)
- [ ] added relevant [`godoc`](https://go.dev/blog/godoc) and [code comments](https://blog.jbowen.dev/2019/09/the-magic-of-go-comments/).
- [ ] updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)

______

### Reviewers Checklist

**All** items are required.
Please add a note if the item is not applicable
and please add your handle next to the items reviewed
if you only reviewed selected items.

I have...

- [ ] confirmed the correct
      [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
      in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] confirmed that this PR does not change production code

-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced oracle price submissions with robust timestamp validation and an improved price hashing mechanism for more accurate aggregation.
  - Introduced native token versioning and support for additional Ethereum chain variants.
  - Staker query responses now include version information for better tracking of validator list changes.
  - Added a new caching mechanism for managing recent messages and parameters in the oracle system.
  - Introduced a new method for clearing all validator report information and missed round data.

- **Improvements**
  - Updated token parameters (e.g., “ETH” now uses 8 decimals) for clearer value representation.
  - Strengthened operator and delegation validations to prevent actions on frozen operators.
  - Updated CI workflows and testing for increased system reliability.
  - Streamlined parameter retrieval processes for enhanced efficiency in price management.
  - Enhanced error handling and logging for better traceability of price submission processes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

leonz789 committed:
use base64 of hash to emit nst-balance-change-event, don't reset index in when nst rounds grows

leonz789 committed:
don't slash on oracle downtime

leonz789 committed:
use lowercase to compare nst prefix

leonz789 committed:
fix: update roundID in staker's balance list

leonz789 committed:
ref:feedermanager. todo:events, uts, aggregator-tally

leonz789 committed:
remove unused file

leonz789 committed:
fix:pepareround,logs. complete:round tally related functions,state copy for simulation

leonz789 committed:
feat: add recovery check in beginblock for devmode-build

leonz789 committed:
feat: update snapshot of feedermanager as state for checkTx mode

leonz789 committed:
emit events in endblock for updated feeders

leonz789 committed:
revert changes to GrowRoundID

leonz789 committed:
add unit tests for aggregation

leonz789 committed:
add equal methods for recovery logic check

leonz789 committed:
remove outdated aggregator

leonz789 committed:
add build tag devmode and corresponding beginblock for recovery check

leonz789 committed:
fix lint

leonz789 committed:
fix lint

leonz789 committed:
revert localnodesh

leonz789 commented:
@coderabbitai review

leonz789 committed:
fix checking for overflow

leonz789 committed:
move e2e out of ut-cover test from ci

leonz789 committed:
separate test jobs

leonz789 commented on x/oracle/keeper/feedermanagement/algo.go:
```diff
@@ -0,0 +1,132 @@
+			a.decimal = price.Decimal
+			return true
+		}
+		if a.decimal != price.Decimal {
+			if a.decimal > price.Decimal {
+				price.Price += strings.Repeat("0", int(a.decimal-price.Decimal))
+				priceInt, _ = new(big.Int).SetString(price.Price, 10)
+			} else {
+				delta := big.NewInt(int64(price.Decimal - a.decimal))
+				for _, v := range a.list {
+					nv := new(big.Int).Mul(v, new(big.Int).Exp(big.NewInt(10), delta, nil))
+					*v = *nv
+				}
+				a.decimal = price.Decimal
+			}
```
This will cause negative number of zeros

leonz789 commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,363 @@
+
+import (
+	"math/big"
+	"reflect"
```
The usage of 'reflect' is for 'devmode' only to test recovery logic

leonz789 commented on x/oracle/keeper/feedermanagement/types.go:
```diff
@@ -0,0 +1,235 @@
+		}
+	}
+	return true
+}
```
removed `Equal`

leonz789 commented on x/oracle/keeper/feedermanagement/round.go:
```diff
@@ -0,0 +1,251 @@
+}
+
+func (r *round) getMsgItemFromProto(msg *oracletypes.MsgItem) *MsgItem {
+	power, _ := r.cache.GetPowerForValidator(msg.Validator)
```
updated

leonz789 commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,363 @@
+	validators := make(map[string]*big.Int)
+	for v, p := range c.validators.validators {
+		validators[v] = new(big.Int).Set(p)
+	}
+	ret.validators = &cacheValidator{
+		validators: validators,
```
map copy, order does not matter.

leonz789 replied:
map copy is safe to range

leonz789 commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,363 @@
+		c.validators.add(item)
+	default:
+		panic("no other types are support")
+	}
```
updated

leonz789 commented on x/oracle/keeper/feedermanagement/helper_test.go:
```diff
@@ -0,0 +1,111 @@
+		ret.records = append(ret.records, t.NewPricePower())
+	}
+	return nil
+}
```
fixed

leonz789 commented on x/oracle/types/params.go:
```diff
@@ -553,6 +555,7 @@ func (p Params) IsValidSource(sourceID uint64) bool {
+		// #nosec G115  // index of array is uint
 		if uint64(k) == feederID {
 			return v
 		}
```
updated

leonz789 commented on x/oracle/keeper/recent_params.go:
```diff
@@ -77,3 +100,42 @@ func (k Keeper) GetAllRecentParamsAsMap(ctx sdk.Context) (result map[int64]*type
+		}
+	}
+	return recentParamsList, prev, latest
+}
```
maxNonce is the length limited to keep for recent Params when set

leonz789 committed:
fix map iteration, ref

leonz789 committed:
print log instead of panic in endBlock

leonz789 committed:
set expected height in query to get deterministic when do e2e oracle test, embedded aggAlgorithm, validate msg

leonz789 committed:
fix lint

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager_test.go:
```diff
@@ -0,0 +1,101 @@
+			fm2.sortedFeederIDs.add(1)
+			fm2.rounds[1].a.ds.AddPriceSource(&ps2, big.NewInt(1), "v1")
+
+			So(fm.Equals(&fm2), ShouldBeTrue)
```
test case. we set specific fields later

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,352 @@
+
+func GetPriceInfoFromProtoPriceTimeDetID(p *oracletypes.PriceTimeDetID) *PriceInfo {
+	return (*PriceInfo)(p)
+}
```
updated

leonz789 commented on x/oracle/keeper/feedermanagement/types.go:
```diff
@@ -0,0 +1,226 @@
+	GetTotalPower() (totalPower *big.Int)
+	GetValidators() []string
+	IsRuleV1(feederID int64) bool
+	IsDeterministic(sournceID int64) (bool, error)
```
fixed

leonz789 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,484 @@
+	if biggestDetID {
+		rds.records = append(rds.records, p)
+	}
+}
```
add comment. The input must be filtered by recordsValidators to make sure no duplicated detID

leonz789 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,484 @@
+	}
+	return pSource.prices, true
+}
+
+func (rv *recordsValidators) GetFinalPrice(algo AggAlgorithm) (*PriceResult, bool) {
+	if rv.finalPrice != nil {
```
in `GetFinalPrice` the keys are well sorted for deterministic

leonz789 replied:
there's no `range` in `GetValidatorQuotePricesForSourceID`

leonz789 commented on x/oracle/keeper/feedermanagement/round.go:
```diff
@@ -0,0 +1,262 @@
+		return
+	}
+	delta = currentHeight - r.startBaseBlock
+	rounds := delta / r.interval
```
updated

leonz789 committed:
refactor

leonz789 commented on .github/workflows/test.yml:
```diff
@@ -44,3 +44,21 @@ jobs:
+  test-unit-e2e:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-go@v4
+        with:
+          go-version: '1.21'
+          check-latest: true
```
upgrade in another PR

leonz789 committed:
documentation

leonz789 commented on x/oracle/keeper/msg_server_create_price.go:
```diff
@@ -24,78 +27,66 @@ func (ms msgServer) CreatePrice(goCtx context.Context, msg *types.MsgCreatePrice
+		if len(priceStr) >= 32 {
+			hash := sha256.New()
+			hash.Write([]byte(priceStr))
+			priceStr = base64.StdEncoding.EncodeToString(hash.Sum(nil))
```
added documentation

leonz789 committed:
refactor

leonz789 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,486 @@
+		return rv.finalPrices, true
+	}
+	ret := make(map[string]*PriceResult)
+	for validator, pv := range rv.records {
+		finalPrice, ok := pv.GetFinalPrice(algo)
+		if !ok {
+			return nil, false
+		}
+		ret[validator] = finalPrice
+	}
```
This process expected a 'all-or-nothing' result, order does not matter

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,361 @@
+
+func (p *PriceInfo) ProtoPriceTimeDetID() *oracletypes.PriceTimeDetID {
+	return (*oracletypes.PriceTimeDetID)(p)
+}
```
updated

leonz789 commented on x/oracle/keeper/feedermanagement/round.go:
```diff
@@ -0,0 +1,265 @@
+		r.status != r2.status {
+		return false
+	}
+	if !r.a.Equals(r2.a) {
+		return false
+	}
```
`q.Equals` handled the `nil` case

leonz789 commented on x/oracle/keeper/feedermanagement/types.go:
```diff
@@ -0,0 +1,226 @@
+type recordsDS struct {
+	finalPrice *PriceResult
+	// TODO: remove this
+	finalDetID        string
```
keep this. remove doc

MaxMustermann2 commented on app/app.go:
```diff
@@ -915,14 +915,15 @@ func NewExocoreApp(
-		stakingtypes.ModuleName,    // track historical info
-		ibcexported.ModuleName,     // handles upgrades of chain and hence client
-		authz.ModuleName,           // clear expired approvals
+		oracleTypes.ModuleName,
```
Please add a comment to explain the positioning.

leonz789 replied:
updated

MaxMustermann2 commented on testutil/utils.go:
```diff
@@ -504,6 +504,7 @@ func (suite *BaseTestSuite) DoSetupTest() {
 	queryHelperEvm := baseapp.NewQueryServerTestHelper(suite.Ctx, suite.App.InterfaceRegistry())
 	evmtypes.RegisterQueryServer(queryHelperEvm, suite.App.EvmKeeper)
 	suite.QueryClientEVM = evmtypes.NewQueryClient(queryHelperEvm)
+	suite.App.OracleKeeper.FeederManager.BeginBlock(suite.Ctx)
```
Why does this need to be called separately ?

leonz789 replied:
not necessary. removed

MaxMustermann2 commented on x/oracle/keeper/cache/caches.go:
```diff
@@ -69,6 +70,7 @@ func (c cacheMsgs) commit(ctx sdk.Context, k common.KeeperOracle) {
 	for ; i < len(index.Index); i++ {
 		b := index.Index[i]
+		// #nosec G115  // maxNonce must not be negative
 		if b > block-uint64(common.MaxNonce) {
```
Check if `block >= uint64(common.MaxNonce)` ?

leonz789 replied:
this package. `cache` is not used. removed this package

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,486 @@
+import (
+	"fmt"
+	"math/big"
+	"reflect"
```
Please check this note about `import "reflect"`

> Certain system packages contain functions which may be a possible source of non-determinism

leonz789 replied:
`reflect` is used in `Equals`, and all the `Equals` defined are used for `devmode` build flag to test recovery logic, so the performance is not that important. Any other concern ?

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,486 @@
+		return false
+	}
+
+	if !reflect.DeepEqual(a.finalPrice, a2.finalPrice) {
```
Instead of reflect, please consider implementing an `Equals` on this `struct`

leonz789 replied:
All `Equals` are used in devemod for test. And why it should be replaced, any concern ?

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,486 @@
+
+func (rv *recordsValidators) RecordMsg(msg *MsgItem) (*MsgItem, error) {
+	record, ok := rv.records[msg.Validator]
+	rets := &MsgItem{
```
nit: Please init this after `if !ok {}`.

leonz789 replied:
updated

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,486 @@
+	}
+	if prices, ok := rv.GetFinalPriceForValidators(algo); ok {
+		keySlice := make([]string, 0, len(prices))
+		for validator := range prices {
```
Please add a comment above this `for` to indicate that the iteration is safe to do because we sort the keys immediately thereafter.

leonz789 replied:
updated

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,486 @@
+		if rv.finalPrice == nil {
+			return nil, false
+		}
+		return rv.finalPrice, true
```
```suggestion
		return rv.finalPrice, rv.finalPrice != nil
```

leonz789 replied:
updated

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,486 @@
+		return rv.finalPrices, true
+	}
+	ret := make(map[string]*PriceResult)
+	for validator, pv := range rv.records {
```
Please add a comment to indicate that this non-deterministic ordered map is consumed only by `GetFinalPrice` which sorts the result by keys.

leonz789 replied:
no, the keys are not sorted. it's not necessary to make this order deterministic. this method will only return true after all items got the 'finalPrice', so the order does not matter.
added a comment

MaxMustermann2 reviewed (commented):
Partial review; asked for more comments next to functions

bwhour commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,486 @@
+}
+
+// TODO: V2: the accumulatedPower should corresponding to all valid validators which provides all sources required by rules(defined in oracle.Params)
+func (a *aggregator) exceedPowerLimit() bool {
```
please take care of the follwoing  **“Ieration over map”** issues

leonz789 replied:
replied

bwhour commented on x/oracle/keeper/feedermanagement/types.go:
```diff
@@ -0,0 +1,225 @@
+type recordsValidators struct {
+	finalPrice  *PriceResult
+	finalPrices map[string]*PriceResult
+	// TODO: V2: accumulatedValidPower only includes validators who prividing all sources required by rules(defined in oracle.Params)
```
typo: prividing

leonz789 replied:
fixed

leonz789 committed:
add version to sync nst validatorlist, fix aggregation for multiple prices with same detID

leonz789 committed:
test: wait one more block to make sure state is committed

leonz789 committed:
remove slashing params for oracle downtime

leonz789 committed:
lint proto

leonz789 commented on proto/exocore/oracle/v1/query.proto:
```diff
@@ -120,8 +120,10 @@
 // QueryStakerListResponse is response type for Query/StakerList RPC method
 message QueryStakerListResponse {
+  // version of the staker validator list changes
+  int64 version = 1;
   // staker list including all stakers of request asset
-  StakerList staker_list = 1;
+  StakerList staker_list = 2;
```
either keep the field number or not, this is a breaking change for cosmossdk module

leonz789 commented on proto/exocore/oracle/v1/query.proto:
```diff
@@ -134,8 +136,10 @@
 // QueryStakerInfoResponse is response type for Query/StakerInfo RCP method
 message QueryStakerInfoResponse {
+  // version of the staker validator list changes
+  int64 version = 1;
   // all staker infos under the specified asset
-  StakerInfo staker_info = 1;
+  StakerInfo staker_info = 2;
```
either keep the field number or not, this is a breaking change for cosmossdk module

leonz789 commented on proto/exocore/oracle/v1/query.proto:
```diff
@@ -146,8 +150,10 @@
+  int64 version = 1;
   // all staker infos under the specified asset
-  repeated StakerInfo staker_infos = 1;
+  repeated StakerInfo staker_infos = 2;
```
either keep the field number or not, this is a breaking change for cosmossdk module

leonz789 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,493 @@
+import (
+	"fmt"
+	"math/big"
+	"reflect"
```
used for devemod in test. And see no reason for 'source of non-deterministic`

leonz789 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,482 @@
+func (a *aggregator) RecordMsg(msg *MsgItem) error {
+	// TODO: implement me
+	_, err := a.v.RecordMsg(msg)
+	return err
```
it's implemented, removed the comment

leonz789 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,482 @@
+		return false
+	}
+	// it's safe to range map here, order does not matter
+	for _, record := range rv.records {
+		// ignore the fail cases for updating some pv' DS finalPrice
+		record.UpdateFinalPriceForDS(sourceID, finalPrice)
+	}
```
No it won't cause no non-deterministic.
Here we update are final prices for corresponding sourceIDs, the order does not matter

leonz789 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,482 @@
+	}
+}
+
+func (rds *recordsDS) GetFinalPrice(t *threshold) (*PriceResult, bool) {
+	if rds.finalPrice != nil {
+		return rds.finalPrice, true
+	}
+	if t.Exceeds(rds.accumulatedPowers) {
+		l := len(rds.records)
+		for i := l - 1; i >= 0; i-- {
+			pPower := rds.records[i]
+			if t.Exceeds(pPower.Power) {
+				rds.finalPrice = pPower.Price.PriceResult()
+				rds.finalDetID = pPower.Price.DetID
+				return rds.finalPrice, true
+			}
+		}
+	}
+	return nil, false
+}
```
`recordsDS` use `AddPrice` to add record in an ordered way. safe.

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,896 @@
+		f.updateCheckTx()
+	}
+	return
+}
```
it's expected to panic, since the error happened in recovery means something wrong in store on node restart. the starting process should be prevented.
moved the panic from recovery out to this BeginBlock to make it clear.

leonz789 committed:
remove package cache, refactor

leonz789 committed:
remove unnecessary condition check

leonz789 committed:
use sorted keys for logs consistance

leonz789 committed:
comments for types

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,895 @@
+	if len(feederIDs) == 0 {
+		return
+	}
+	validators := f.cs.GetValidators()
```
It's safe to range map here since this is for memory update, anyway it's updated to use sorted keys for consistent logs display.

leonz789 committed:
refactor:remove unused code

leonz789 committed:
revert sorted keys and use map range directly for nonce remove, sort keys before nonce insert

leonz789 committed:
remove redundant init for validators in endblock, remove function duplicated for init

leonz789 committed:
hanld quote behavior in deterministic order with sorted keys, add comments to explain why map iteration is safe

leonz789 committed:
test: add an e2e case for quoting after delegation

leonz789 commented on testutil/network/genesis_data.go:
```diff
@@ -40,8 +40,8 @@ var (
 		Tokens: []assetstypes.StakingAssetInfo{
-			NewTestToken("ETH", "Ethereum native token", ETHAssetAddress, TestEVMChainID, 5000),
-			NewTestToken("NST ETH", "native restaking ETH", NativeAssetAddress, TestEVMChainID, 5000),
+			NewTestToken("ETH", "Ethereum native token", ETHAssetAddress, TestEVMChainID, 0, 5000),
+			NewTestToken("NST ETH", "native restaking ETH", NativeAssetAddress, TestEVMChainID, 0, 5000),
```
It's expected for usage in oracle test cases

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,909 @@
+			// items will be removed from slice and keep the order, so it's safe to delete items in different order
+			f.k.RemoveNonceWithFeederIDForAll(ctx, uint64(r.feederID))
+		}
+	}
```
It's safe to range maps here and no need to hash every keys to found values.

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,896 @@
+
+func (f *FeederManager) ValidateMsg(msg *oracletypes.MsgCreatePrice) error {
+	// TODO: implement me
+	// nonce, feederID, creator has been verified by anteHandler
```
this method has been implemented, remove the comment

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,829 @@
+// initCaches initializes the caches of the FeederManager with keeper, params, validatorPowers
+func (f *FeederManager) initCaches(ctx sdk.Context) {
+	f.cs = newCaches()
+	params := f.k.GetParams(ctx)
+	validatorSet := f.k.GetAllExocoreValidators(ctx)
+	validatorPowers := make(map[string]*big.Int)
+	for _, v := range validatorSet {
+		validatorPowers[sdk.ConsAddress(v.Address).String()] = big.NewInt(v.Power)
+	}
```
the returned order of validator list from `GetAllExcoreValidators(ctx)` is deterministic

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,895 @@
+	if len(f.rounds) != len(fm.rounds) {
+		return false
+	}
+	for id, r := range f.rounds {
+		if r2, ok := fm.rounds[id]; !ok {
+			return false
+		} else if !r.Equals(r2) {
+			return false
+		}
+	}
```
Equals requires all kvs in map are the same or return false if any is not equal. So the order does not matter.

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,369 @@
+	if ps.sourceID != ps2.sourceID || ps.deterministic != ps2.deterministic {
+		return false
+	}
+	if !reflect.DeepEqual(ps.detIDs, ps2.detIDs) {
```
1. This is used for devmode
2. i don't see any non-deterministic concen for the compare case, any detail ?

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,336 @@
+		return ps.prices[i].DetID < ps.prices[j].DetID
+	})
+	return ret, nil
+}
```
It's actually safe for this scenario here since this is only for internal calculation. However we can change it to SliceStable for consistent semantics

leonz789 commented on x/oracle/keeper/recent_msg.go:
```diff
@@ -6,6 +6,29 @@ import (
+		for ; i < len(index.Index); i++ {
+			b := index.Index[i]
+			// #nosec G115  // maxNonce is not negative
+			if b > block-uint64(maxNonce) {
```
fixed

leonz789 commented on x/oracle/types/key_slashing.go:
```diff
@@ -6,15 +6,23 @@ var (
+	return MissedBitArrayPrefix
+}
+
+func SlashingMissedBitArrayValidatorPrefix(validator string) []byte {
 	key := append([]byte(validator), DelimiterForCombinedKey)
 	return append(MissedBitArrayPrefix, key...)
 }
```
it's defined in types. moved from var to constant

leonz789 commented on x/oracle/types/types.go:
```diff
@@ -43,17 +44,54 @@ type AggFinalPrice struct {
 	return valueBytes
 }
+
+func ConsAddrStrFromCreator(creator string) (string, error) {
+	accAddress, err := sdk.AccAddressFromBech32(creator)
+	if err != nil {
+		return "", err
+	}
+	return sdk.ConsAddress(accAddress).String(), nil
+}
```
this fail is included in the next check: `sdk.AccAddressFromBech32(creator)`

leonz789 committed:
refactor

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,895 @@
+}
+
+// 1. update and commit Params if updated
+// 2. update and commit validatorPowers if updated
```
safe to range map. order does not matter, update memory state only.

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,895 @@
+					logger.Info("commit round with aggregated price", "feederID", r.feederID, "roundID", r.roundID, "baseBlock", r.roundBaseBlock, "price", priceCommit, "heigth", height)
+
+					// #nosec G115  // tokenID is index of slice
+					f.k.AppendPriceTR(ctx, uint64(r.tokenID), *priceCommit, finalPrice.DetID)
```
safe to range map. no kvstore will be updated in recovery mode

leonz789 committed:
refactor

fix lint
remove unused code

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,376 @@
+package feedermanagement
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"reflect"
+	"slices"
+	"sort"
+
+	oracletypes "github.com/ExocoreNetwork/exocore/x/oracle/types"
+)
```
Those are used for devmode

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,376 @@
+
+func (p *PriceResult) PriceInfo() *PriceInfo {
+	return (*PriceInfo)(p)
+}
```
updated

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,921 @@
+
+	f.updateCheckTx()
+
+}
```
we don't expect no errors in those methods

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,921 @@
+		f.cs.Read(params)
+		existsFeederIDs := make(map[int64]struct{})
+		// safe to range map. update memory state only, the result would be the same in any order
+		for _, r := range f.rounds {
+			r.UpdateParams(params.TokenFeeders[r.feederID], int64(params.MaxNonce))
+			existsFeederIDs[r.feederID] = struct{}{}
+		}
```
safe to range map here, only values for keys matter, the order does not impact the result

leonz789 commented on x/oracle/keeper/validator_update_block.go:
```diff
@@ -6,6 +6,12 @@ import (
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.ValidatorUpdateBlockKey))
+	b := k.cdc.MustMarshal(&validatorUpdateBlock)
+	store.Set(types.BlockKey, b)
+}
```
removed this method

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,895 @@
+		recovered, err = f.recovery(ctx)
+		// it's safe to panic since this will only happen when the node is starting with something wrong in the store
+		if err != nil {
+			panic(err)
```
expected. this will only panic on node restart when something wrong with db state

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,895 @@
+		recovered, err = f.recovery(ctx)
+		// it's safe to panic since this will only happen when the node is starting with something wrong in the store
+		if err != nil {
+			panic(err)
```
expected. this will only panic on node restart when something wrong with db state

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,909 @@
+func (f *FeederManager) EndBlockInRecovery(ctx sdk.Context, params *oracletypes.Params) {
+	if params != nil {
+		f.SetParamsUpdated()
+		_ = f.cs.AddCache(params)
+	}
+	f.updateAndCommitRoundsInRecovery(ctx)
+	f.prepareRounds(ctx)
+	f.ResetFlags()
+}
+
+func (f *FeederManager) setupNonces(ctx sdk.Context, feederIDs []int64) {
+	logger := f.k.Logger(ctx)
+	height := ctx.BlockHeight()
+	// the order does not matter, it's safe to update independent state in non-deterministic order
+	// no need to go through all 'hash' process to range sorted key slice
+	for _, r := range f.rounds {
+		// remove nonces for closed quoting windows or when forceSeal is marked
+		if r.IsQuotingWindowEnd(height) || f.forceSeal {
+			logger.Debug("clear nonces for closing quoting window or forceSeal", "feederID", r.feederID, "roundID", r.roundID, "basedBlock", r.roundBaseBlock, "height", height, "forceSeal", f.forceSeal)
+			// #nosec G115  // feederID is index of slice
+			// items will be removed from slice and keep the order, so it's safe to delete items in different order
+			f.k.RemoveNonceWithFeederIDForAll(ctx, uint64(r.feederID))
+		}
+	}
```
safe.
duplicated alert.

leonz789 commented on x/oracle/module_beginblock.go:
```diff
@@ -0,0 +1,13 @@
+
+// BeginBlock contains the logic that is automatically triggered at the beginning of each block
+func (am AppModule) BeginBlock(ctx sdk.Context, _ abci.RequestBeginBlock) {
+	am.keeper.BeginBlock(ctx)
```
safe.
duplicated alert.

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,369 @@
+	if p == nil || p2 == nil {
+		return false
+	}
+	if !reflect.DeepEqual(p.Price, p2.Price) || p.Power.Cmp(p2.Power) != 0 {
+		return false
+	}
+	if len(p.Validators) != len(p2.Validators) {
+		return false
```
safe.
duplicated alert.

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,369 @@
+	"errors"
+	"fmt"
+	"math/big"
+	"reflect"
```
safe.
duplicated alert.

leonz789 replied:
safe.
duplicated alert.

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,895 @@
+
+// initCaches initializes the caches of the FeederManager with keeper, params, validatorPowers
+func (f *FeederManager) initCaches(ctx sdk.Context) {
+	f.cs = newCaches()
+	params := f.k.GetParams(ctx)
+	validatorSet := f.k.GetAllExocoreValidators(ctx)
+	validatorPowers := make(map[string]*big.Int)
+	for _, v := range validatorSet {
+		validatorPowers[sdk.ConsAddress(v.Address).String()] = big.NewInt(v.Power)
+	}
```
safe.
duplicated alert.

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,369 @@
+	if len(updated) > 0 {
+		return updated, added, nil
+	}
+	return nil, nil, fmt.Errorf("failed to add priceSource listi, error:%s", es)
+}
+
```
safe.
duplicated alert.

leonz789 committed:
refactor nil compare for Equals

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,392 @@
+
+	if !ps.deterministic {
+		// this is not ds, then just set the final price or overwrite if the input has a later timestamp
+		if ps.finalPrice == nil {
+			ps.finalPrice = psNew.prices[0].PriceResult()
+			ps.prices = append(ps.prices, psNew.prices[0])
+			psNew.prices = psNew.prices[:1]
+			return psNew, nil
+		}
+		// equivalent to After, just overwrite the old value
+		if psNew.prices[0].Timestamp > ps.finalPrice.Timestamp {
+			ps.finalPrice = psNew.prices[0].PriceResult()
+			ps.prices = append(ps.prices, psNew.prices[0])
+			psNew.prices = psNew.prices[:1]
+			return ps, nil
```
updated

leonz789 commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,392 @@
+	}
+}
+
+func (p *PriceInfo) EqualDS(pi *PriceInfo) bool {
+	return p.Price == pi.Price && p.DetID == pi.DetID && p.Decimal == pi.Decimal
+}
```
fixed

leonz789 committed:
typo

bwhour commented on Dockerfile:
```diff
@@ -23,7 +23,14 @@ WORKDIR /root
 COPY --from=build-env /go/bin/toml-cli /usr/bin/toml-cli
 
-RUN apk add --no-cache ca-certificates=20241121-r1 libstdc++=13.2.1_git20231014-r0 jq=1.7.1-r0 curl=8.11.1-r0 bash=5.2.21-r0 vim=9.0.2127-r0 lz4=1.9.4-r5 rclone=1.65.0-r3 \
+RUN apk add --no-cache ca-certificates=20241121-r1 \
+	libstdc++~=13.2 \
+	jq~=1.7 \
+	curl~=8.11 \
+	bash~=5.2 \
+	vim~=9.0 \
+	lz4~=1.9 \
+	rclone~=1.65 \
```
good point.

leonz789 replied:
removed in #291

bwhour commented on testutil/network/genesis_data.go:
```diff
@@ -80,10 +80,10 @@ func init() {
 	DefaultGenStateOracle.Params.Slashing.ReportedRoundsWindow = 4
 	// set jailduration of oracle report downtime to 30 seconds for test
-	DefaultGenStateOracle.Params.Slashing.OracleMissJailDuration = 30 * time.Second
+	DefaultGenStateOracle.Params.Slashing.OracleMissJailDuration = 15 * time.Second
```
Also change the commentto 15’

leonz789 replied:
updated

bwhour commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+			f.SetForceSeal()
+		}
+		if paramsOld.IsSlashingResetUpdate(&params) {
+			f.SetResetSlasing()
```
typo: resetSlasing

leonz789 replied:
it's expected to be 'SetResetSlashing', the flag is 'resetSlashing'

bwhour commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+	f.validatorsUpdated = true
+}
+
+func (f *FeederManager) SetResetSlasing() {
```
typo: SetResetSlasing

leonz789 replied:
it's expected to be 'SetResetSlashing', the flag is 'resetSlashing'

bwhour commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,487 @@
+	}
+}
+
+// type recordsDSs struct {
```
remove this comment.

leonz789 replied:
removed

bwhour commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,376 @@
+// RuleV1, we restrict the source to be Chainlink and only that source is acceptable
+func (c *caches) IsRuleV1(feederID int64) bool {
+	ruleID := c.params.params.TokenFeeders[feederID].RuleID
+	return ruleID == 1 && len(c.params.params.Sources) == 2 && c.params.params.Sources[1].Name == oracletypes.SourceChainlinkName
```
please change the magic number to const definition.

leonz789 replied:
updated

bwhour commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,376 @@
+func (c *caches) GetTokenIDForFeederID(feederID int64) (int64, bool) {
+	tf, ok := c.GetTokenFeederForFeederID(feederID)
+	if !ok {
+		return 0, false
```
change magic number 0

leonz789 replied:
it comes as 0 for int with return false value bool

bwhour commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,376 @@
+	c.params.update = false
+}
+
+// CommitCache commits the cache to the KVStore
```
typo: the CommitCache function name

leonz789 replied:
updated

leonz789 committed:
add e2e and ut cases, add copy algo for fchecktx

leonz789 committed:
refactor, typo

leonz789 commented on tests/e2e/oracle/recovery.go:
```diff
@@ -0,0 +1,458 @@
+	res, err = s.network.QueryOracle().LatestPrice(ctxWithHeight(start+3), &oracletypes.QueryGetLatestPriceRequest{TokenId: 1})
+	s.Require().NoError(err)
+	// price not updated yet
+	s.Require().Equal(res.Price.Price, priceRecovery3.Prices[0].Price)
+
+	//    2.2 mixed with some different prices(detID, price)
+	// init_start+110
+	start += 10
+	startUint = uint64(start)
+	s.moveToAndCheck(start)
+	msg3_2.BasedBlock = startUint
+	msg1_2.BasedBlock = startUint
+	msg2_2.BasedBlock = startUint
+	msg0_2.BasedBlock = startUint
+	// msg0_2.BasedBlock = startUint
+	// id:3
+	err = s.network.SendTxOracleCreateprice([]sdk.Msg{msg3_2}, "valconsKey3", kr3)
+	s.Require().NoError(err)
+
+	// id:1,2,3
+	err = s.network.SendTxOracleCreateprice([]sdk.Msg{msg1_2}, "valconsKey1", kr1)
+	s.Require().NoError(err)
+
+	s.moveToAndCheck(start + 1)
+
+	// id:1,2
+	err = s.network.SendTxOracleCreateprice([]sdk.Msg{msg2_2}, "valconsKey2", kr2)
+	s.Require().NoError(err)
+
+	// id:3
+	//	err = s.network.SendTxOracleCreateprice([]sdk.Msg{msg0_2}, "valconsKey0", kr0)
+	//	s.Require().NoError(err)
+
+	s.moveToAndCheck(start + 3)
+
+	res, err = s.network.QueryOracle().LatestPrice(ctxWithHeight(start+2), &oracletypes.QueryGetLatestPriceRequest{TokenId: 1})
+	s.Require().NoError(err)
+	// price not updated yet
+	s.Require().Equal(res.Price.Price, priceRecovery3.Prices[0].Price)
+
```
the height when to sending msg and after how many blocks to verify the result are variable

leonz789 committed:
fix nil compare

leonz789 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+		f = f.getCheckTx()
+	}
+	if err := f.ValidateMsg(msg); err != nil {
+		return nil, oracletypes.ErrInvalidMsg.Wrap(err.Error())
+	}
+	msgItem := getProtoMsgItemFromQuote(msg)
+
+	// #nosec G115  // feederID is index of slice
+	r, ok := f.rounds[int64(msgItem.FeederID)]
+	if !ok {
+		// This should not happened since we do check the nonce in anthHandle
+		return nil, fmt.Errorf("round not exists for feederID:%d, proposer:%s", msgItem.FeederID, msgItem.Validator)
+	}
+
+	// #nosec G115  // baseBlock is block height which is not negative
+	if valid := r.ValidQuotingBaseBlock(int64(msg.BasedBlock)); !valid {
+		return nil, fmt.Errorf("failed to process price-feed msg for feederID:%d, round is quoting:%t,quotingWindow is open:%t, expected baseBlock:%d, got baseBlock:%d", msgItem.FeederID, r.IsQuoting(), r.IsQuotingWindowOpen(), r.roundBaseBlock, msg.BasedBlock)
+	}
+
+	// tally msgItem
+	finalPrice, validMsgItem, err := r.Tally(msgItem)
+
+	// record msgItem in caches if needed
+	defer func() {
+		if !isCheckTx &&
+			validMsgItem != nil &&
+			(err == nil || sdkerrors.IsOf(err, oracletypes.ErrQuoteRecorded)) {
+			_ = f.cs.AddCache(validMsgItem)
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+
+	if finalPrice == nil {
+		return nil, nil
+	}
+	return finalPrice.ProtoPriceTimeRound(r.roundID, ctx.BlockTime().Format(oracletypes.TimeLayout)), nil
+}
```
tested in tests/e2e/oracle

leonz789 committed:
refactor
add unused code
add test cases for recovery check in e2e tets, including force seal
add test cases to elaborate all quoting cases in uint-test
keep at least one history recent params when update

leonz789 commented on x/oracle/keeper/feedermanagement/data_test.go:
```diff
@@ -0,0 +1,275 @@
+package feedermanagement
+
+import oracletypes "github.com/ExocoreNetwork/exocore/x/oracle/types"
+
```
it's well defined in helper_test.go

leonz789 commented on x/oracle/keeper/feedermanagement/data_test.go:
```diff
@@ -0,0 +1,275 @@
+			},
+		},
+		Validator: "validator4",
+	}
```
updated

leonz789 commented on x/oracle/keeper/recent_params.go:
```diff
@@ -77,3 +109,40 @@ func (k Keeper) GetAllRecentParamsAsMap(ctx sdk.Context) (result map[int64]*type
+func (k Keeper) GetRecentParamsWithinMaxNonce(ctx sdk.Context) (recentParamsList []*types.RecentParams, prev, latest types.RecentParams) {
+	maxNonce := k.GetParams(ctx).MaxNonce
+	var startHeight uint64
+	if uint64(ctx.BlockHeight()) > uint64(maxNonce) {
+		startHeight = uint64(ctx.BlockHeight()) - uint64(maxNonce)
+	}
```
it's - not +

leonz789 committed:
remove commentted codes, refactor

leonz789 committed:
use math.pow to calculate count

leonz789 committed:
regenerate swagger doc

leonz789 committed:
regenerate proto

leonz789 committed:
typo: fix annatation

leonz789 commented on client/docs/swagger-ui/swagger.json:
```diff
@@ -19907,6 +19897,11 @@
             "schema": {
               "type": "object",
               "properties": {
+                "version": {
+                  "type": "string",
+                  "format": "int64",
+                  "title": "asset id for the staker infos request for"
+                },
```
No, actually the first field is changed to 'version'. fixed the annatation

leonz789 committed:
regenerate proto, swagger

TimmyExogenous commented on x/oracle/keeper/msg_server_create_price.go:
```diff
@@ -24,78 +28,69 @@ func (ms msgServer) CreatePrice(goCtx context.Context, msg *types.MsgCreatePrice
 
-	agc := ms.Keeper.GetAggregatorContext(ctx)
-	newItem, caches, err := agc.NewCreatePrice(ctx, msg)
+	if err := ms.ValidateMsg(msg); err != nil {
```
In `ProcessQuote`, `ValidateMsg` will be called again. So should we remove the validation here?

leonz789 replied:
removed

TimmyExogenous commented on x/oracle/keeper/feedermanagement/types.go:
```diff
@@ -0,0 +1,284 @@
+	status roundStatus
+	// aggregator is the price aggregator for current round
+	a *aggregator
+	// cache is the cache reader for current round to provid params, validators information
```
typo: provide

leonz789 replied:
updated

TimmyExogenous commented on x/oracle/keeper/feedermanagement/types.go:
```diff
@@ -0,0 +1,284 @@
+
+// each source will get a final price independently, the order of sources does not matter, map is safe
+// recordsDSs is the price records for all deterministic sources
+// threshold indicates the threshold definition to detemin final price for each source
```
typo: detemin

leonz789 replied:
updated

cloud8little commented:
@MaxMustermann2 @bwhour could you have a review?

TimmyExogenous commented on x/oracle/keeper/feedermanagement/prices.go:
```diff
@@ -0,0 +1,389 @@
+			return psNew, nil
+		}
+		// equivalent to After, just overwrite the old value
+		if psNew.prices[0].Timestamp > ps.finalPrice.Timestamp {
```
It seems that the logic is the same as `ps.finalPrice == nil`. The only difference is that it returns `ps`, while the previous one returns `psNew`. Could these two cases be merged when addressing this?

leonz789 replied:
updated

TimmyExogenous commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+	// update params and validator set if necessary in caches and commit all updated information
+	addedValidators := f.updateAndCommitCaches(ctx)
+
+	// update Slashing related records (reportInfo, missCountBitArray), handle case for 1. reseetSlashing, 2. new validators added for validatorset change
```
typo: reseetSlashing

leonz789 replied:
updated

TimmyExogenous commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+
+	// set status to open for rounds before their quoting window
+	feederIDs := f.prepareRounds(ctx)
+	// remove nocnes for closing quoting-window and set nonces for opening quoting-window
```
typo: nocnes

leonz789 replied:
updated

TimmyExogenous commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+			r.FinalPrice()
+			r.status = roundStatusClosed
+		}
+		// close all quotingWindow to skip current rounds' 'handlQuotingMisBehavior'
```
typo: should be `handleQuotingMisBehavior`

leonz789 replied:
updated

TimmyExogenous commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,378 @@
+	return
+}
+
+// SkipCommit skip real commit by setting the updage flag to false
```
typo: updage

leonz789 replied:
updated

leonz789 committed:
rebase

adu-web3 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,485 @@
+	if !a.v.Equals(a2.v) {
+		return false
+	}
+	if !a.ds.Equals(a2.ds) {
```
how about `a.algo` vs `a2.algo`, maybe we could add some metadata method like `type` for algo interface so that we could strictly compare the aggregator

leonz789 replied:
fixed

leonz789 committed:
ref, typo

bwhour commented on x/oracle/keeper/keeper.go:
```diff
@@ -36,7 +28,8 @@ type (
 		types.SlashingKeeper
 		// wrap all four memory cache into one pointer to track them among cpoies of Keeper (msgServer, module)
-		memStore *memoryStore
+		// TODO: remove this
+		*feedermanagement.FeederManager
```
@leonz789 pls check this ai comment and resolve.

leonz789 committed:
typo

adu-web3 commented on app/ante/cosmos/sigverify.go:
```diff
@@ -391,6 +391,8 @@ func (isd IncrementSequenceDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, sim
 			if accAddress, err := sdk.AccAddressFromBech32(msg.Creator); err != nil {
 				return ctx, errors.New("invalid address")
+				// #nosec G115  // safe conversion
+				// TODO: define msg.Nonce as uint32 to avoid conversion
```
do we check that `msg.Creator` is a validator before check and increase the nonce?

leonz789 replied:
It's checked in the `CheckAndIncreaseNonce` which makes sure the creator is currently an active validator

adu-web3 commented on x/oracle/keeper/recent_params.go:
```diff
@@ -77,3 +109,40 @@ func (k Keeper) GetAllRecentParamsAsMap(ctx sdk.Context) (result map[int64]*type
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.RecentParams
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		latest = val
```
this functions seems to assume that recent params would be iterated by the order of block from oldest to latest, is this safe?

leonz789 replied:
The key is uint64 of block height convert as BigEndian to bytes. So the lexicographical order would be satisfied.

adu-web3 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+		// which means no memory state need to be updated for thoes rounds
+		// and we don't need to take care of 'close quoting-window' since the size of replay window t most equals to maxNonce
+		// #nosec G115  // block is not negative
+		if len(recentMsgs) > 0 && int64(recentMsgs[0].Block) <= startHeight {
```
is it safe to assume the recentMsgs are ordered? (I suppose this take the first msg as oldest)

leonz789 replied:
It's the same as 'recentParams', the key is convert from block height as uint64 to BigEndian bytes.

adu-web3 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+
+	ctxReplay := ctx.WithBlockHeight(replayHeight)
+	for tfID, tf := range params.TokenFeeders {
+		if tfID == 0 {
```
why the first tokenfeeder is skipped?

leonz789 replied:
We limit all IDs(corresponding to slice index) start from 1.
tokenID, tokenFeederID, sourceID, ruleID...

adu-web3 commented on x/oracle/keeper/feedermanagement/algo.go:
```diff
@@ -0,0 +1,132 @@
+
+func (b BigIntList) Median() *big.Int {
+	sort.Sort(b)
+	l := len(b)
```
should we assert the length to make sure b is not empty?

leonz789 replied:
updated

leonz789 committed:
add Equals for aggAlgorithm

return nil for empty list when do median

adu-web3 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+			logger.Debug("clear nonces for closing quoting window or forceSeal", "feederID", r.feederID, "roundID", r.roundID, "basedBlock", r.roundBaseBlock, "height", height, "forceSeal", f.forceSeal)
+			// #nosec G115  // feederID is index of slice
+			// items will be removed from slice and keep the order, so it's safe to delete items in different order
+			f.k.RemoveNonceWithFeederIDForAll(ctx, uint64(r.feederID))
```
not related to this PR's change, but it seems `oracleKeeper. RemoveNonceWithFeederIDForAll` is implemented in a redundant way, because it involves redundant lops:
```golang
func (k Keeper) RemoveNonceWithFeederIDForAll(ctx sdk.Context, feederID uint64) {
	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.NonceKeyPrefix))
	iterator := store.Iterator(nil, nil)
	defer iterator.Close()
	var validators []string
	for ; iterator.Valid(); iterator.Next() {
		var nonce types.ValidatorNonce
		k.cdc.MustUnmarshal(iterator.Value(), &nonce)
		validators = append(validators, nonce.Validator)
	}
	k.removeNonceWithFeederIDForValidators(store, feederID, validators)
}
```
we could have removed the nonce from each validator's nonce list in one loop

leonz789 replied:
Yes, that could reduce one loop.
the unexported method `removeNonceWithFeederIDForValidators` is extracted as common logic to be invoked by both:
`RemoveNonceWithFeederIDForValidators()`
`RemoveNonceWithFeederIDForAll()`

adu-web3 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+		r := f.rounds[feederID]
+		logger.Debug("init nonces for new quoting window", "feederID", feederID, "roundID", r.roundID, "basedBlock", r.roundBaseBlock, "height", height)
+		// #nosec G115 -- feederID is index of slice
+		f.k.AddZeroNonceItemWithFeederIDForValidators(ctx, uint64(feederID), validators)
```
may not be related to this PR's change, but should we store the nonce with `feederID` as index in `NonceList`, thus we might save extra costs of searching the `NonceList` to find specific `nonce` that matches the `feederID`:
```go
type ValidatorNonce struct {
	// validator is the validator address
	Validator string `protobuf:"bytes,1,opt,name=validator,proto3" json:"validator,omitempty"`
	// nonces is the list of nonces for the feeders
	NonceList []*Nonce `protobuf:"bytes,2,rep,name=nonce_list,json=nonceList,proto3" json:"nonce_list,omitempty"`
}
```

```go
func (k Keeper) AddZeroNonceItemWithFeederIDForValidators(ctx sdk.Context, feederID uint64, validators []string) {
	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.NonceKeyPrefix))
	for _, validator := range validators {
		if n, found := k.getNonce(store, validator); found {
			found := false
			for _, v := range n.NonceList {
				if v.FeederID == feederID {
					found = true
					break
				}
			}
			if !found {
				n.NonceList = append(n.NonceList, &types.Nonce{FeederID: feederID, Value: 0})
				k.setNonce(store, n)
			}
		} else {
			k.setNonce(store, types.ValidatorNonce{Validator: validator, NonceList: []*types.Nonce{{FeederID: feederID, Value: 0}}})
		}
	}
}
```

leonz789 replied:
Yes that's a feasible appraoch.
However the tradeoff is: 
in the 'feederID as index' way, when we have only feederID=9 works, we need to keep 8 empty items in the slice.
as in current approach, it only need to keep one item. Or we can just use KV to keep nonce for validator_feederID :

Read one:
- now
  - 1 time I/O, iterate slice in memory
- list with index for feederID
  - 1 time I/O ( empty iterms for inactive tokenFeeders )
- KV
  - 1 time I/O

Write/update multiple feederIDs:
- now
  - 2 times I/O(read + write), iterate slice in memory
- list
  - 2 times I/O(read + write), iterate slice in memory(set empty items)
- KV
  - 2*N times I/O

Remove All(multiple feederIDs), for each validator:
- now
  - 1 time I/O
- list
  - 1 time I/O
- KV
  - N times I/O

Memory iteration is much cheaper than DB iteration. For performance optimization, we can batch feederIDs update for one validator in one invoke(updated in the latest commit).

adu-web3 commented on x/oracle/keeper/feedermanagement/types.go:
```diff
@@ -0,0 +1,284 @@
+// FeederManager is the manager for the price feed progress of all token feeders
+type FeederManager struct {
+	// fCheckTx is a copy of FeederManager used for mode of checkTx to simulate transactions
+	fCheckTx *FeederManager
```
Do we need this copy of the `FeederManager` itself in `FeederManager` because we do not want to pollute the memory state of `FeederManager` during `CheckTx` phase?

leonz789 replied:
fCheckTx is updated at `EndBlock` only, and it's access through `getCheckTx` which return a copy of fCheckTx instead of itself.

adu-web3 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,917 @@
+	f.cs.Init(f.k, &params, validatorPowers)
+}
+
+func (f *FeederManager) recovery(ctx sdk.Context) (bool, error) {
```
the `recovery` phase of recovering the in-memory state of `FeederManager` is very complex, and seems important to the consistent state of `FeederManager`, is there any doc that I could refer to so that I could get better understanding?

leonz789 replied:
https://www.notion.so/exocore/Oracle-module-recovery-19d612d2b5df80b2b023c1297eda3e06?pvs=4
cc. @MaxMustermann2 
recovery logic explain

bwhour reviewed (approved):
LGTM.

MaxMustermann2 commented on app/app.go:
```diff
@@ -915,14 +915,17 @@ func NewExocoreApp(
-		stakingtypes.ModuleName,    // track historical info
-		ibcexported.ModuleName,     // handles upgrades of chain and hence client
-		authz.ModuleName,           // clear expired approvals
+		// being block of oracle will fill params cache, need to be put before epochs will use the params
```
```suggestion
		// begin block of oracle will fill params cache, need to be put before epochs will use the params
```

Why is epochs using the params?

leonz789 replied:
hooks.AfterEpochEnd -> operator.UpdateVotingPower -> oracle.GetMultipleAssetsPrices -> k.GetParamsFromCache

MaxMustermann2 commented on proto/exocore/oracle/v1/params.proto:
```diff
@@ -71,14 +71,8 @@ message SlashingParams {
-  ];
   // slash_fraction_malicious defines the fraction one validator should be punished for malicious behavior
-  bytes slash_fraction_malicious = 6 [
+  bytes slash_fraction_malicious = 5 [
```
Is this different from the double sign x/slashing parameter?

leonz789 replied:
The value can be the same, however it might be better to define the params independently as part of oracle since this malicious check is based on oracle 'fake price' instead of 'double sign'

MaxMustermann2 replied:
I will defer to you but I don't think we will keep the number different for the two cases. But yes, it's good to have this parameter.

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,378 @@
+
+type ItemV map[string]*big.Int
+
+var zeroBig = big.NewInt(0)
```
`common.Big0` is also nice but not a blocker

leonz789 replied:
updated

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,378 @@
+				delete(cv.validators, operator)
+				cv.update = true
+			} else if power.Cmp(newPower) != 0 {
+				cv.validators[operator].Set(newPower)
```
Is it always updated to the new power?

leonz789 replied:
yes, this `add` works the same as `k.ApplyValidatorChanges` in dogfood.
refactored to make the logic clean

leonz789 committed:
refactor, remove unused interface

leonz789 committed:
refactor

leonz789 committed:
refactor remove unnecessary check

leonz789 committed:
Revert "refactor"

This reverts commit 7fc9beff02cfdf5cae9ff727e956affb879d1a60.

leonz789 commented on x/oracle/keeper/nonce.go:
```diff
@@ -151,26 +69,80 @@ func (k Keeper) removeNonceWithValidator(store prefix.Store, validator string) {
+			}
+		}
 	}
 }
```
we use map to filter the removed items, newList will result to different orders

leonz789 committed:
comments

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/aggregator.go:
```diff
@@ -0,0 +1,491 @@
+}
+
+// TODO: V2: the accumulatedPower should corresponding to all valid validators which provides all sources required by rules(defined in oracle.Params)
+func (a *aggregator) exceedPowerLimit() bool {
```
This function name is confusing. Is it threshold > power ?

leonz789 replied:
yes, renamed to `exceedesThreshold`

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/algo.go:
```diff
@@ -0,0 +1,152 @@
+	notNumber
+)
+
+type AggMedian struct {
```
Add an interface guard for this to ensure it implements `AggAlgorithm`.

leonz789 replied:
added

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/caches.go:
```diff
@@ -0,0 +1,380 @@
+	return
+}
+
+// GetTotalPower returns the total power of all validators
```
This is the power accumulated so far, right?

leonz789 replied:
yes, updated `cachevalidators.add` to make it clean and also avoid redundant calculation.

leonz789 committed:
Revert "update docker dependency"

This reverts commit 131e85e1681eb432aebc994c077e4a1062b166c1.

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/feedermanager.go:
```diff
@@ -0,0 +1,933 @@
+	validatorUpdateBlock, found := f.k.GetValidatorUpdateBlock(ctx)
+	if !found {
+		// on recovery mode, the validator update block must be found, otherwise we just panic to stop the node start
+		// it's safe to panic since this will only happen when the node is starting with something wrong in the store
```
The line following this comment is not a `panic`.

leonz789 replied:
If 'error' returned from recovery the `BeginBlock` will do panic.
removed the 'panic' part in this comment

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/round.go:
```diff
@@ -0,0 +1,263 @@
+}
+
+func (r *round) ValidQuotingBaseBlock(height int64) bool {
+	return r.IsQuotingWindowOpen() && r.roundBaseBlock == height
```
Why `==` ?

leonz789 replied:
We use 'baseblock' to sync the message update price.
For round start from 11, its baseBlock is 10, all messages sent to update price for this round should set 'baseBlock' to 10 in that message to tell it's quoting for that round from baseBlock-10.

MaxMustermann2 commented on x/oracle/keeper/feedermanagement/round.go:
```diff
@@ -0,0 +1,263 @@
+	price.DetID = detID
+	prices, ok := r.a.v.GetValidatorQuotePricesForSourceID(validator, oracletypes.SourceChainlinkID)
+	if !ok {
+		return
```
Should `miss` be `true` here?

leonz789 replied:
yes `miss` is set to `true` in line 217, to cover cases not matched in line 225-232(which should also be 'true'). However it's not clear, refactored this method to make it clean in the latest commit

MaxMustermann2 commented on x/oracle/keeper/msg_server_create_price.go:
```diff
@@ -24,78 +28,65 @@ func (ms msgServer) CreatePrice(goCtx context.Context, msg *types.MsgCreatePrice
-	logger.Info("add price proposal for aggregation", "feederID", msg.FeederID, "basedBlock", msg.BasedBlock, "proposer", msg.Creator, "height", ctx.BlockHeight())
-
+	logger.Info("added quote for aggregation", append(logQuote, "msg", msg)...)
+	// TODO: use another type
```
Please update for indexer.

MaxMustermann2 commented on x/oracle/keeper/msg_server_create_price.go:
```diff
@@ -24,78 +28,65 @@ func (ms msgServer) CreatePrice(goCtx context.Context, msg *types.MsgCreatePrice
+		if len(priceStr) >= maxPriceLength {
+			hash := sha256.New()
+			hash.Write([]byte(priceStr))
+			priceStr = base64.StdEncoding.EncodeToString(hash.Sum(nil))
```
The indexer will not be able to figure out the price from this. Any other ideas?

leonz789 replied:
This is used for 'nst balance change' which might be of too big size for event.
For 'price' it will be of the original size.
Maybe 'ignore' the hash case (length >=32bytes), or just display the hash ?

MaxMustermann2 replied:
Can you split the attribute into different parts? One for decimal, one for price, one for tokenID and one for roundID ?

MaxMustermann2 commented on x/oracle/keeper/native_token.go:
```diff
@@ -333,6 +349,47 @@ func (k Keeper) UpdateNSTByBalanceChange(ctx sdk.Context, assetID string, rawDat
+	value := store.Get(key)
+	if value == nil {
+		// set the first index of version to 1
+		store.Set(key, k.cdc.MustMarshal(&types.NSTVersion{AssetId: assetID, Version: 1}))
```
Redundant storage of `assetID` in the `value` when it is already present in the key. Instead confisder using `sdk.Uint64ToBigEndian(version)`

leonz789 replied:
updated

leonz789 commented on networks/local/exocore/Dockerfile:
```diff
@@ -12,7 +12,7 @@ RUN LEDGER_ENABLED=false make build
 RUN apk add --no-cache libstdc++~=13.2 \
 	bash~=5.2 \
-	curl~=8.11 \
+	curl~=8.12 \
```
the file is updated to downgrade that version to 8.11

leonz789 committed:
use big0 from ethereum/common

typo

rename exceedPowerLimit to exceedsThreshold

interface guard for aggMedian

refactor cachevalidators.add, cache totalpower to avoid calculation

refactor PerformanceReview

remove nstversion, use uint64bytes instead

leonz789 committed:
skip querying version when error not nil

return false when input is nil

leonz789 commented on x/oracle/keeper/query_native_token.go:
```diff
@@ -24,7 +24,11 @@
+
+	stakerInfosResp, err := k.GetStakerInfos(ctx, req)
+	version := k.GetNSTVersion(ctx, req.AssetId)
+	stakerInfosResp.Version = version
```
fixed

leonz789 commented on x/oracle/keeper/slashing.go:
```diff
@@ -26,6 +26,16 @@ func (k Keeper) InitValidatorReportInfo(ctx sdk.Context, validator string, heigh
+		store.Delete(iterator.Key())
+	}
+	iterator.Close()
+}
```
called separately

leonz789 commented on x/oracle/types/params.go:
```diff
@@ -594,3 +607,22 @@ func (p Params) CheckDecimal(feederID uint64, decimal int32) bool {
+		return true
+	}
+	return false
+}
```
updated

leonz789 committed:
upgrade curl version

leonz789 committed:
change exocore to imuachain in semgrep config
